File: player-core/src/main/java/com/libreshockwave/player/score/SpriteSpan.java
package com.libreshockwave.player.score;

import com.libreshockwave.id.ChannelId;
import com.libreshockwave.id.FrameId;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents a span of frames where a sprite/behavior is active.
 * Channel 0 is reserved for frame behaviors, channels 1+ are for sprites.
 */
public class SpriteSpan {

    private final ChannelId channel;
    private final FrameId startFrame;  // Inclusive
    private final FrameId endFrame;    // Inclusive
    private final List<ScoreBehaviorRef> behaviors;

    public SpriteSpan(int channel, int startFrame, int endFrame) {
        this.channel = new ChannelId(channel);
        // Score data can have startFrame/endFrame of 0; clamp to 1 for FrameId validity
        this.startFrame = new FrameId(Math.max(1, startFrame));
        this.endFrame = new FrameId(Math.max(1, endFrame));
        this.behaviors = new ArrayList<>();
    }

    public int getChannel() {
        return channel.value();
    }

    public ChannelId getChannelId() {
        return channel;
    }

    public int getStartFrame() {
        return startFrame.value();
    }

    public FrameId getStartFrameId() {
        return startFrame;
    }

    public int getEndFrame() {
        return endFrame.value();
    }

    public FrameId getEndFrameId() {
        return endFrame;
    }

    public List<ScoreBehaviorRef> getBehaviors() {
        return behaviors;
    }

    public void addBehavior(ScoreBehaviorRef behavior) {
        behaviors.add(behavior);
    }

    /**
     * Check if this span is a frame behavior (channel 0).
     */
    public boolean isFrameBehavior() {
        return channel.value() == 0;
    }

    /**
     * Check if the given frame is within this span's range.
     */
    public boolean containsFrame(int frame) {
        return frame >= startFrame.value() && frame <= endFrame.value();
    }

    /**
     * Get the first behavior in this span (typically the main behavior).
     */
    public ScoreBehaviorRef getFirstBehavior() {
        return behaviors.isEmpty() ? null : behaviors.get(0);
    }

    @Override
    public String toString() {
        return "SpriteSpan{channel=" + channel.value() +
               ", frames=" + startFrame.value() + "-" + endFrame.value() +
               ", behaviors=" + behaviors.size() + "}";
    }
}

File: cpp/src/player/score/SpriteSpan.cpp
#include "libreshockwave/player/score/SpriteSpan.hpp"

#include <algorithm>
#include <sstream>
#include <utility>

namespace libreshockwave::player::score {

SpriteSpan::SpriteSpan(int channel, int startFrame, int endFrame)
    : channel_(channel),
      startFrame_(std::max(1, startFrame)),
      endFrame_(std::max(1, endFrame)) {}

int SpriteSpan::channel() const { return channel_.value(); }
id::ChannelId SpriteSpan::channelId() const { return channel_; }
int SpriteSpan::startFrame() const { return startFrame_.value(); }
id::FrameId SpriteSpan::startFrameId() const { return startFrame_; }
int SpriteSpan::endFrame() const { return endFrame_.value(); }
id::FrameId SpriteSpan::endFrameId() const { return endFrame_; }
const std::vector<ScoreBehaviorRef>& SpriteSpan::behaviors() const { return behaviors_; }

void SpriteSpan::addBehavior(ScoreBehaviorRef behavior) {
    behaviors_.push_back(std::move(behavior));
}

bool SpriteSpan::isFrameBehavior() const {
    return channel_.value() == 0;
}

bool SpriteSpan::containsFrame(int frame) const {
    return frame >= startFrame_.value() && frame <= endFrame_.value();
}

const ScoreBehaviorRef* SpriteSpan::firstBehavior() const {
    return behaviors_.empty() ? nullptr : &behaviors_.front();
}

std::string SpriteSpan::toString() const {
    std::ostringstream out;
    out << "SpriteSpan{channel=" << channel_.value()
        << ", frames=" << startFrame_.value() << "-" << endFrame_.value()
        << ", behaviors=" << behaviors_.size() << "}";
    return out.str();
}

} // namespace libreshockwave::player::score
